pkg/kubelet/dockertools/docker_manager.go

```diff
@@ -597,15 +614,22 @@ func (dm *DockerManager) runContainer(
 		cpuShares = milliCPUToShares(cpuRequest.MilliValue())
 	}
 	var devices []dockercontainer.DeviceMapping
+	var nvidiaGPUPaths []string
+	nvidiaGPUPlugin := dm.GetNvidiaGPUPlugin()
 	if nvidiaGPULimit.Value() != 0 {
-		// Experimental. For now, we hardcode /dev/nvidia0 no matter what the user asks for
-		// (we only support one device per node).
-		devices = []dockercontainer.DeviceMapping{
-			{"/dev/nvidia0", "/dev/nvidia0", "mrw"},
-			{"/dev/nvidiactl", "/dev/nvidiactl", "mrw"},
-			{"/dev/nvidia-uvm", "/dev/nvidia-uvm", "mrw"},
+		if nvidiaGPUPlugin != nil {
+			nvidiaGPUPaths = nvidiaGPUPlugin.AllocateGPU(int(nvidiaGPULimit.Value()))
+			if nvidiaGPUPaths != nil {
+				devices = []dockercontainer.DeviceMapping{{PathOnHost: gpuutil.NvidiaDeviceCtl, PathInContainer: gpuutil.NvidiaDeviceCtl, CgroupPermissions: "mrw"},
+					{PathOnHost: gpuutil.NvidiaDeviceUVM, PathInContainer: gpuutil.NvidiaDeviceUVM, CgroupPermissions: "mrw"}}
+
+				for _, path := range nvidiaGPUPaths {
+					devices = append(devices, dockercontainer.DeviceMapping{PathOnHost: path, PathInContainer: path, CgroupPermissions: "mrw"})
+				}
+			}
```

Comment:
Is there a way we can move this logic into kubelet? The less runtime specific code the better. I'm envisioning a mechanism where the kubelet "assigns" a GPU to a container, and then here you just check for the assignment and setup the appropriate docker command in that case.